parsl/executors/high_throughput/process_worker_pool.py

```diff
@@ -312,6 +335,14 @@ def pull_tasks(self, kill_event):
                 self.heartbeat_to_incoming()
                 last_beat = time.time()
 
+            if self.drain_time and time.time() > self.drain_time:
+                logger.info("Requesting drain")
+                self.drain_to_incoming()
+                self.drain_time = None
+                # This will start the pool draining...
+                # Drained exit behaviour does not happen here. It will be
+                # driven by the interchange sending a DRAINED_CODE message.
```

Comment:
Any thoughts on passing `drain_period` to the interchange and moving this logic there to avoid passing messages back-and-forth?

Reply:
The original structure of the "please drain me now" (workers->interchange) pattern comes from a potential extension that is you don't know a drain period ahead of time, but get told that by your batch system as a signal - so that the "please drain me" message would be requested due to a Mysterious External Event rather than a timer (you tell slurm: "give me a signal a few minutes before my batch job ends").

In the end I didn't implement that but I think I'd like to keep that as an implementation possibility to support a couple of potential use cases: submitting batch jobs with different characteristics (see issue #3113) and supporting pre-emptive queues (where a user submits a batch job with "I don't care how long I run for, just tell me before you terminate me" in exchange for some other valuable resource - for example 50% charging discount at NERSC in the `preempt` queue - https://docs.nersc.gov/jobs/policy/)

I think there has to be a "you are drained, you can exit now" message (interchange->workers) because that's how the worker pool knows it can exit. It can't use a "exit when empty" decision because because a pool will often be empty: for example, a 1 worker pool will be empty between every single task.

Reply:
That works for me.